### PR TITLE
CME reg v2 WIP

### PR DESCRIPTION
**Pull request by sigberto**
@stvstnfrd @caesar2164 
This is still a work in progress, but so you can get an idea of where it's going. The only major thing left is the javascript stuff we were talking about today—I'll update at least some of it this weekend.

### CODE REVIEW

**Comment by stvstnfrd** [comment]
@sigberto Thanks for pushing this. Could you also add a checklist that details the outstanding work here? Specifically, what behavior on the form still needs addressed?

**Comment by sigberto** [comment]
added most points on the trello card, but there is some stuff outside of edx-platform that needs to change too:
Basically, look at http://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/customize_registration_page.html points 3 onward

**Comment by sigberto** [comment]
Change lms.env.json to reflect this:

"ADDL_INSTALLED_APPS": [
        "class_dashboard",
        "cme_reg_form"
],

"REGISTRATION_EXTENSION_FORM": "cme_reg_form.forms.ExtraInfoForm",

**Comment by dcadams** [comment]
General comment on the migrations. Can you squash them into a single migration file rather than the several. You should be able to do this by deleting all the migration files and then running the auto generate migrations again.

File: common/djangoapps/cme_registration_v2/cme_reg_form/admin.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+from django.contrib import admin
+from .models import ExtraInfo
+
+admin.site.register(ExtraInfo)
```
Comment (dcadams) [inline common/djangoapps/cme_registration_v2/cme_reg_form/admin.py]: Do we need this file (admin.py).

File: common/djangoapps/cme_registration_v2/cme_reg_form/models.py [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+                                        ('MD,PhD', 'MD,PhD'),
+                                        ('MBBS', 'MBBS'),
+                                        ('NP', 'NP'),
+                                        ('PA', 'PA'),
+                                        ('PharmD', 'PharmD'),
+                                        ('PhD', 'PhD'),
+                                        ('RN', 'RN'),
+                                        ('Other', 'Other'),
+                                        ('None', 'None'))
+    professional_designation = models.CharField(max_length=10, choices=PROFESSIONAL_DESIGNATION_CHOICES)
```
Comment (dcadams) [inline common/djangoapps/cme_registration_v2/cme_reg_form/models.py]: change to max_length 25

File: common/djangoapps/cme_registration_v2/cme_reg_form/models.py [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+                          ('Structural Biology Department', 'Structural Biology Department'),
+                          ('Surgery', 'Surgery'),
+                          ('Urology', 'Urology'),
+                          ('Urology - Administration', 'Urology - Administration'),
+                          ('Urology - Divisions', 'Urology - Divisions'))
+    stanford_department = models.CharField(max_length=50, blank=True, choices=DEPARTMENT_CHOICES)
+
+    sunet_id = models.CharField(max_length=33, blank=True)
+
+    address_1 = models.CharField(max_length=255)
```
Comment (dcadams) [inline common/djangoapps/cme_registration_v2/cme_reg_form/models.py]: Change to TextField.

File: common/djangoapps/cme_registration_v2/cme_reg_form/models.py [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+    address_2 = models.CharField(max_length=255, blank=True)
```
Comment (dcadams) [inline common/djangoapps/cme_registration_v2/cme_reg_form/models.py]: Change to TextField.

File: common/djangoapps/cme_registration_v2/cme_reg_form/models.py [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+    city = models.CharField(max_length=50, blank=True)
```
Comment (dcadams) [inline common/djangoapps/cme_registration_v2/cme_reg_form/models.py]: Change to TextField.

File: common/djangoapps/cme_registration_v2/cme_reg_form/models.py [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+    sub_specialty = models.CharField(max_length=255, blank=True)
+
+    AFFILIATION_CHOICES = (('Stanford Children\'s Health', 'Stanford Children\'s Health'),
+                           ('Packard Children\'s Health Alliance (PCHA)', 'Packard Children\'s Health Alliance (PCHA)'),
+                           ('Stanford Health Care', 'Stanford Health Care'),
+                           ('Stanford University', 'Stanford University'),
+                           ('University Healthcare Alliance (UHA)', 'University Healthcare Alliance (UHA)'),
+                           ('Not affiliated with Stanford Medicine', 'Not affiliated with Stanford Medicine'),
+                           )
+    affiliation = models.CharField(max_length=50, choices=AFFILIATION_CHOICES)
```
Comment (dcadams) [inline common/djangoapps/cme_registration_v2/cme_reg_form/models.py]: Change to max_length=46

File: common/djangoapps/cme_registration_v2/cme_reg_form/models.py [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+    other_affiliation = models.CharField(max_length=50, blank=True) # Not used?
```
Comment (dcadams) [inline common/djangoapps/cme_registration_v2/cme_reg_form/models.py]: Change to max_length=46

File: common/djangoapps/cme_registration_v2/cme_reg_form/models.py [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+
+    sub_affiliation = models.CharField(max_length=50, blank=True)
```
Comment (dcadams) [inline common/djangoapps/cme_registration_v2/cme_reg_form/models.py]: Change to max_length=46

File: common/djangoapps/cme_registration_v2/cme_reg_form/models.py [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+                          ('SoM Dean\'s Office Administrative Units', 'SoM Dean\'s Office Administrative Units'),
+                          ('SoM Non Cap Projects', 'SoM Non Cap Projects'),
+                          ('Stanford Cancer/Stem Cell Biology', 'Stanford Cancer/Stem Cell Biology'),
+                          ('Stanford Institutes of Medicine', 'Stanford Institutes of Medicine'),
+                          ('Structural Biology Department', 'Structural Biology Department'),
+                          ('Surgery', 'Surgery'),
+                          ('Urology', 'Urology'),
+                          ('Urology - Administration', 'Urology - Administration'),
+                          ('Urology - Divisions', 'Urology - Divisions'))
+    stanford_department = models.CharField(max_length=50, blank=True, choices=DEPARTMENT_CHOICES)
```
Comment (dcadams) [inline common/djangoapps/cme_registration_v2/cme_reg_form/models.py]: Change to max_length=46

**Comment by dcadams** [comment]
All the columns in cme_reg_form_extrainfo are null not allowed. For a nullable input field, what gets put into the table?

**Comment by dcadams** [comment]
We need a data migration to migrate the existing data to the new schema.

**Comment by sigberto** [comment]
@dcadams steven is working on a PR on top of this one here:
https://github.com/Stanford-Online/edx-platform/pull/619
so comments may more appropriate there. @stvstnfrd not sure if you got the notifications.

**Comment by sigberto** [comment]
@dcadams "null" on a field means that the field is set to NULL in the database. however, specifically for string fields, django stores empty answers as empty strings. they recommend not setting null to true in that case. the "blank" condition specifies whether a field can be empty. see https://docs.djangoproject.com/en/1.8/ref/models/fields/#field-options for more detail

File: common/djangoapps/cme_registration_v2/cme_reg_form/models.py (hunk 9 above)
Comment (sigberto) [inline common/djangoapps/cme_registration_v2/cme_reg_form/models.py]: i'm sure there is one, but what's the reason for being so specific about the charfield lengths? is it just consistency with the current schema?

File: common/djangoapps/cme_registration_v2/cme_reg_form/models.py (hunk 5 above)
Comment (sigberto) [inline common/djangoapps/cme_registration_v2/cme_reg_form/models.py]: same here, is the reason for textfield just consistency? for short inputs, the database is more efficient with charfield than textfield: http://stackoverflow.com/questions/7354588/django-charfield-vs-textfield/8698298

File: common/djangoapps/cme_registration_v2/cme_reg_form/admin.py (hunk 1 above)
Comment (sigberto) [inline common/djangoapps/cme_registration_v2/cme_reg_form/admin.py]: honestly, not sure. it was used in the examples i followed.

**Comment by dcadams** [comment]
Well, unfortunately, there are some records with null values for fields in the existing table. These look like they were created during a testing phase but they'll need special consideration for the data migration.

File: common/djangoapps/cme_registration_v2/cme_reg_form/admin.py (hunk 1 above)
Comment (dcadams) [inline common/djangoapps/cme_registration_v2/cme_reg_form/admin.py]: I think it is only required if we need to fiddle with the model in django admin which I don't think we will.

**Comment by stvstnfrd** [comment]
Closing in favor of #619

**Pull request closed without merging**